torchdata/stateful_dataloader/worker.py

```diff
@@ -213,10 +213,13 @@ def _worker_loop(
                                 _DATASET_ITER_STATE: try_to_serialize(fetcher.dataset_iter),  # type: ignore[union-attr]
                                 _FETCHER_ENDED: fetcher.ended,  # type: ignore[union-attr]
                             }
+                            # Pick up any user-defined dataset state if it is not the iterator as it is already captured in fetcher_state's dataset_iter_state
+                            dataset_state = try_to_serialize(dataset) if fetcher.dataset_iter is not dataset else None  # type: ignore[union-attr]
```

Comment:
nice

Reply:
Question: when they are the same, why choose iterable over dataset's state?

Reply:
dataset state works too honestly, I just picked one. I can switch it to dataset state as it is one level higher in the nesting dict hierarchy?

Reply:
I can't quite articulate why but it seems like dataset state seems more correct, I can't think of a breaking example yet though